org.osgi.test.common/src/main/java/org/osgi/test/common/install/InstallBundle.java

```diff
@@ -18,21 +18,294 @@
+			}
+
+			findBundle(bundleContext, bundleSymbolicName, bundleVersion).orElseThrow(() -> {
+				return new IllegalArgumentException(String.format(
+					"Unknown Bundle with SymbolicName '%s' and Version: '%s'", bundleSymbolicName, bundleVersion));
+			});
+			return new EmbeddedLocation(bundleSymbolicName, bundleVersion, oPath.orElse("/"), file);
+		}
+
+		/**
+		 * Creates an EmbeddedLocation by the spec. Existence of File is tested
+		 * at creation time.
+		 *
+		 * @param bundleContext
+		 * @param spec 'bundle.symbilic.name:1.2.3:/path/file.jar'
+		 *            <p>
+		 *            In minimum 'file.jar' must be used. Then it searches the
+		 *            file in the root of the BundelContexts bundle.
+		 * @return EmbeddedLocation
+		 */
+		public static EmbeddedLocation of(BundleContext bundleContext, String spec) {
+
+			Optional<String> oBundleSymbolicName;
+			Optional<String> oBundleVersion;
+			String[] parts = spec.split(":");
+			String fullPath;
+			if (parts.length == 1) {
+
+				oBundleSymbolicName = Optional.empty();
+				oBundleVersion = Optional.empty();
+				fullPath = parts[0];
+			} else if (parts.length == 2) {
+				oBundleSymbolicName = Optional.of(parts[0]);
+				oBundleVersion = Optional.empty();
+				fullPath = parts[1];
+			} else if (parts.length == 3) {
+				oBundleSymbolicName = Optional.of(parts[0]);
+				oBundleVersion = Optional.of(parts[1]);
+				fullPath = parts[2];
+			} else {
+				throw new IllegalArgumentException("Wrong pattern :" + spec);
+			}
+
+			int indexOfFileBegin = fullPath.lastIndexOf('/') + 1;
+
+			String file = fullPath.substring(indexOfFileBegin);
+
+			if (file.isEmpty()) {
+				throw new IllegalArgumentException(String.format("could not find a filename : ", spec));
+
+			}
+			Optional<String> oPath = Optional.of(fullPath.substring(0, indexOfFileBegin));
+
+			return EmbeddedLocation.of(bundleContext, oBundleSymbolicName, oBundleVersion, oPath, file);
+
+		}
+
+		/**
+		 * Creates an EmbeddedLocation by the pure parameters. existence of File
+		 * not tested at creation time.
+		 *
+		 * @param bundleSymbolicName - symbolicName of the bundle that embedded
+		 *            the file must not be null
+		 * @param bundleVersion - version of the bundle that embedded the file -
+		 *            must not be null
+		 * @param path - the path in the jar - must not be null
+		 * @param file - the filename must not be null
+		 * @return EmbeddedLocation
+		 */
+		public static EmbeddedLocation of(String bundleSymbolicName, Version bundleVersion, String path, String file) {
+			return new EmbeddedLocation(bundleSymbolicName, bundleVersion, path, file);
+		}
+
+		private String	bundleSymbolicName;
+
+		private Version	bundleVersion;
+
+		private String	file;
+		private String	path;
+
+		private EmbeddedLocation() {
+			// no access here.
+		}
+
+		private EmbeddedLocation(String bundleSymbolicName, Version bundleVersion, String path, String file) {
+			if (bundleSymbolicName == null) {
```

Comment:
These should be `Objects.requireNonNull` tests.

Reply:
now use Objects methods